### ISSUE TITLE
Securing Graph API Requests

### ISSUE THREAD

**Issue by blackhpro** [yiisoft/yii2-authclient#67]
Hi.
I do not know whether there was raised the question but I want to ask it.

When using Facebook ```php yii\authclient\clients\Facebook``` sometimes an error occurs" api calls from the server require an appsecret_proof argument ". It occurs on facebook if not passed mandatory parameter appsecret_proof.
If you enable additional security check.
Here it is written in more detail - https://developers.facebook.com/docs/graph-api/securing-requests

Fix it pretty easy, you have already done in such class as the LinkedIn / VKontakte. It is enough to override a method Facebook :: apiInternal
adding

```php
protected function apiInternal($accessToken, $url, $method, array $params, array $headers)
{                
    $params['appsecret_proof'] = hash_hmac('sha256', $accessToken->getToken(), $this->clientSecret);        
    return $this->sendRequest($method, $url, $params, $headers);
}
```

Is it possible to use the safety appsecret_proof without making any changes to class oAuth client?

**Comment by klimov-paul** [comment]
If particular request requires extra parameters you should build and pass them on your own.

**Comment by SDKiller** [comment]
As for now `"API calls from the server require an appsecret_proof argument"` error is produced even with default `email` scope.
This should be fixed "oficially", IMO.

FYI, 
https://github.com/thephpleague/oauth2-facebook/issues/17
https://github.com/thephpleague/oauth2-facebook/pull/18

**Comment by klimov-paul** [comment]
> As for now "API calls from the server require an appsecret_proof argument" error is produced even with default email scope.

Unable to reproduce. Default configuration of the `Facebook` client works fine wih current code, while it includes `email` scope.

**Comment by SDKiller** [comment]
I received this error yesterday each time with 2 fresh created facebook apps:

1. Install `AuthChoice` widget
2. Configure `AuthAction` in controller
3. In `successCallback` call `$client->getUserAttributes()` (to perform some preparations before register/login user)

Error has gone after overriding Facebook client with patch implemented:
```
    public function applyAccessTokenToRequest($request, $accessToken)
    {
        $data = $request->getData();
        $data['access_token'] = $accessToken->getToken();
        $data['appsecret_proof'] = hash_hmac('sha256', $accessToken->getToken(), $this->clientSecret);
        $request->setData($data);
    }
```

**Comment by klimov-paul** [comment]
According to the docs: there is a option `Require App Secret`, which if enabled require `appsecret_proof` for all API requests.
It seems the changes become necessary here.

**Comment by klimov-paul** [comment]
Resolved by commit 2b08ef174f5d194fd2cbb5e6b1a66383f1466905
